README.md
# VScode Web Preview
⚠️ WARNING: This extension is still under development! ⚠️

An extension for you to preview your web projects on!

This is not a live-server it's live-preview. You can use any local server like http://localhost:3000. You can change url.


## Keyboard Shortcuts

`ctrl+shift+p` command **Open Web Preview** and **Web Preview Url**

`ctrl+alt+p` **Open Web Preview** and `ctrl+alt+u` open **Web Preview Url** input box. Open **Web Preview URL** when web preview is active.

### Live preview in vscode

<video src="https://raw.githubusercontent.com/jabed-dev/vscode-webpreview/main/demo.mp4" autoplay loop controls muted title="Live web preview in vscode"></video>

<br>

Set default screen size and url in settings.json (example)
```json
    "webPreview.url": "http://localhost:3000",
    "webPreview.mediaScreenOverride": true, // Override default screen object
    "webPreview.mediaScreen": {
        "Mobile": "380x796", // width and height
        "Tablet": 768,       // width
        "Laptop": 1024,      // width
    }
```

<br>

Creator: **Jabed Hossain**\
https://github.com/jabed-dev

![creator](https://avatars.githubusercontent.com/u/59810760?s=180&v=4)

CHANGELOG.md
# Change Log

All notable changes to the "preview" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

react/src/App.js
import { useState, useRef } from 'react'
import { Preview } from './components/Preview'

function App() {
  const previewRef = useRef()
    , [responsiveDesignMode, setResponsiveDesignMode] = useState(false)
    , DEFAULT_RESPONSIVE_SIZE = { width: 380, height: (window.innerHeight - 41) }
    , [responsiveSize, setResponsiveSize] = useState(DEFAULT_RESPONSIVE_SIZE)
    , [url, setUrl] = useState('http://localhost:3000')
    , [mediaScreens, setMediaScreens] = useState({})
    , media = window.matchMedia('(min-width: 500px)').matches
    
  // Handle the message inside the webview
  window.addEventListener('message', (event) => {
    const { preview } = event.data;
    if (preview) {
      if (preview.url) {
        setUrl(preview.url)
      }
      if (preview.mediaScreen) {
        setMediaScreens(preview.mediaScreen)
      }
      if (preview.responsive) {
        setResponsiveDesignMode(media && !responsiveDesignMode)
      }
      if (preview.refresh) {
        try {
          setUrl(new URL(url))
        } catch {
          setUrl(url)
        }
      }
    }
  });

  return (
    <main className="w-full h-full">
      <Preview
        ref={previewRef}
        frameUrl={url}
        mediaScreen={mediaScreens}
        responsiveDesignMode={media && responsiveDesignMode}
        responsiveSize={responsiveSize}
        onChangeResponsiveSize={setResponsiveSize}
      />
    </main>
  )
}

export default App;


src/extension.ts
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
	const onOpen = vscode.commands.registerCommand('preview.open', () => {
		PreviewPanel.createOrShow(context.extensionUri);
	});
	const onUrl = vscode.commands.registerCommand('preview.url', async () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.sendUrl();
		}
	});
	const onBack = vscode.commands.registerCommand('preview.back', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.sendPageBack();
		}
	});
	const onForward = vscode.commands.registerCommand('preview.forward', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.sendPageForward();
		}
	});
	const onRefresh = vscode.commands.registerCommand('preview.refresh', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.sendRefresh();
		}
	});
	const onResponsiveView = vscode.commands.registerCommand('preview.responsiveView', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.sendResponsiveView()
		}
	});
	const onScreenView = vscode.commands.registerCommand('preview.screenView', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.toggleScreenView()
		}
	});
	const onOpenDevTools = vscode.commands.registerCommand('preview.openDevTools', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.openDevTools()
		}
	});
	const onOpenInBrowser = vscode.commands.registerCommand('preview.openInBrowser', () => {
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel.openInBrowser()
		}
	});

	context.subscriptions.push(onOpen, onUrl, onBack, onForward, onRefresh, onResponsiveView, onScreenView, onOpenDevTools, onOpenInBrowser);

	if (vscode.window.registerWebviewPanelSerializer) {
		// Make sure we register a serializer in activation event
		vscode.window.registerWebviewPanelSerializer(PreviewPanel.viewType, {
			async deserializeWebviewPanel(webviewPanel: vscode.WebviewPanel, state: any) {
				// Reset the webview options so we use latest uri for `localResourceRoots`.
				webviewPanel.webview.options = getWebviewOptions(context.extensionUri);
				PreviewPanel.revive(webviewPanel, context.extensionUri, state?.previewUrl);
			}
		});
	}
}

function getWebviewOptions(extensionUri: vscode.Uri): vscode.WebviewOptions {
	return {
		// Enable javascript in the webview
		enableScripts: true,
		// And restrict the webview to only loading content from our extension's `media` directory.
		localResourceRoots: [vscode.Uri.joinPath(extensionUri, 'media')]
	};
}

/**
 * Manages Preview webview panels
 */
class PreviewPanel {
	/**
	 * Track the currently panel. Only allow a single panel to exist at a time.
	 */
	public static currentPanel: PreviewPanel | undefined;
	public static readonly viewType = 'preview';
	private readonly _panel: vscode.WebviewPanel;
	private readonly _extensionUri: vscode.Uri;
	private _disposables: vscode.Disposable[] = [];
	public _currentUrl: string | undefined;

	public static createOrShow(extensionUri: vscode.Uri) {
		// If we already have a panel, show it.
		if (PreviewPanel.currentPanel) {
			PreviewPanel.currentPanel._panel.reveal(PreviewPanel.currentPanel._panel.viewColumn);
			return;
		}
		// Otherwise, create a new panel.
		const panel = vscode.window.createWebviewPanel(
			PreviewPanel.viewType,
			'Preview',
			vscode.ViewColumn.Two,
			getWebviewOptions(extensionUri),
		);

		PreviewPanel.currentPanel = new PreviewPanel(panel, extensionUri);
	}

	public static revive(panel: vscode.WebviewPanel, extensionUri: vscode.Uri, oldUrl: string) {
		PreviewPanel.currentPanel = new PreviewPanel(panel, extensionUri);
		PreviewPanel.currentPanel._currentUrl = oldUrl;
	}

	private constructor(panel: vscode.WebviewPanel, extensionUri: vscode.Uri) {
		this._panel = panel;
		this._extensionUri = extensionUri;
		vscode.commands.executeCommand('setContext', 'PreviewActive', true);
		this._panel.webview.html = this._getHtmlForWebview(this._panel.webview);
		panel.iconPath = {
			light: vscode.Uri.joinPath(extensionUri, 'media', 'icons', 'preview-light.svg'),
			dark: vscode.Uri.joinPath(extensionUri, 'media', 'icons', 'preview-dark.svg'),
		};
		// Set the webview's initial html content
		this._update();
		// Listen for when the panel is disposed
		// This happens when the user closes the panel or when the panel is closed programmatically
		this._panel.onDidDispose(() => this.dispose(), null, this._disposables);

		// Update the content based on view changes
		this._panel.onDidChangeViewState(e => {
			if (this._panel.visible) {
				this._update();
			}
		},
			null,
			this._disposables
		);

		vscode.workspace.onDidChangeConfiguration(e => {
			if (e.affectsConfiguration('webPreview')) {
				this._currentUrl = undefined;
				this._update()
			}
		})

		// Handle messages from the webview
		this._panel.webview.onDidReceiveMessage(message => {
			if (message.command) {
				vscode.window.showInformationMessage(message.text);
				return;
			}
		},
			null,
			this._disposables
		);
	}

	public async sendUrl() {
		const url = await vscode.window.showInputBox({
			value: this._currentUrl,
			// @ts-ignore
			valueSelection: [this._currentUrl?.length, this._currentUrl?.length],
			placeHolder: 'http://localhost:3000'
		});
		if (!url) return;
		this._currentUrl = url;
		this._panel.webview.postMessage({ preview: { url } });
	}

	public sendPageBack() {
		this._panel.webview.postMessage({ preview: { back: true } });
	}

	public sendPageForward() {
		this._panel.webview.postMessage({ preview: { forward: true } });
	}

	public sendRefresh() {
		this._panel.webview.postMessage({ preview: { refresh: true } });
	}
	
	public sendResponsiveView() {
		this._panel.webview.postMessage({ preview: { responsive: true } });
	}
	
	public toggleScreenView() {
		// @ts-ignore
		if (this._panel.viewColumn > 1) {
			this._panel.reveal(vscode.ViewColumn.One);
		} else if (this._panel.viewColumn == 1) {
			this._panel.reveal(vscode.ViewColumn.Two);
		}
	}

	public openDevTools() {
		vscode.commands.executeCommand('workbench.action.webview.openDeveloperTools')
	}
	
	public openInBrowser() {
		const url = vscode.Uri.parse(`${this._currentUrl}`);
		vscode.env.openExternal(url);
	}

	public dispose() {
		PreviewPanel.currentPanel = undefined;
		vscode.commands.executeCommand('setContext', 'PreviewActive', false)
		// Clean up our resources
		this._panel.dispose();

		while (this._disposables.length) {
			const x = this._disposables.pop();
			if (x) {
				x.dispose();
			}
		}
	}

	private _update() {
		const config = vscode.workspace.getConfiguration('webPreview');
		const url = config.get('url', 'http://localhost:3000');
		if (!this._currentUrl) {
			this._currentUrl = url;
		}

		const mediaScreenOverride = config.get('mediaScreenOverride', false);
		let mediaScreen = config.get('mediaScreen', {});
		if (mediaScreenOverride) {
			const inspect = config.inspect('mediaScreen');
			// @ts-ignore
			mediaScreen = inspect.workspaceValue ? inspect.workspaceValue : inspect.globalValue ? inspect.globalValue : inspect.defaultValue
		}
		// Vary the webview's content based on where it is located in the editor.
		if (this._panel.viewColumn) {
			this._panel.webview.postMessage({ preview: { url: this._currentUrl, mediaScreen } });
		}
	}

	private _getHtmlForWebview(webview: vscode.Webview) {
		// Local path to main script run in the webview
		const scriptPathOnDisk = vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js');
		// And the uri we use to load this script in the webview
		const scriptUri = webview.asWebviewUri(scriptPathOnDisk);
		// Local path to css styles
		const stylesPathMainPath = vscode.Uri.joinPath(this._extensionUri, 'media', 'main.css');
		// Uri to load styles into webview
		const stylesMainUri = webview.asWebviewUri(stylesPathMainPath);
		// Use a nonce to only allow specific scripts to be run
		const nonce = getNonce();

		return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
                <meta name="theme-color" content="#000000">
				<title>Preview</title>
				<link href="${stylesMainUri}" rel="stylesheet">
                <script defer="defer" nonce="${nonce}" src="${scriptUri}"></script>
			</head>
			<body class="w-full min-h-full flex text-white bg-gray-900 !p-0 overflow-hidden">
                <noscript>You need to enable JavaScript to run this app.</noscript>
                <div id="root" class="w-full flex flex-none flex-col"></div>
            </body>
			</html>`;
	}
}

function getNonce() {
	let text = '';
	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < 32; i++) {
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	}
	return text;
}


react/src/components/Preview.js
import { forwardRef, useEffect, useState, useRef, useLayoutEffect } from 'react'
import { ResponsiveSizing } from './ResponsiveSizing'

function getPointerPosition(event) {
  if (event.targetTouches) {
    if (event.targetTouches.length === 1) {
      return {
        x: event.targetTouches[0].clientX,
        y: event.targetTouches[0].clientY,
      }
    }
    return null
  }
  return { x: event.clientX, y: event.clientY }
}

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect
export const Preview = forwardRef(
  (
    {
      responsiveDesignMode,
      responsiveSize,
      onChangeResponsiveSize,
      frameUrl,
      mediaScreen
    },
    ref
  ) => {
    const containerRef = useRef()
    const [size, setSize] = useState({ width: 0, height: 0 })
    const [resizing, setResizing] = useState()
    const timeout = useRef()
    const constrainedResponsiveSize = constrainSize(
      responsiveSize.width,
      responsiveSize.height
    )

    function constrainWidth(desiredWidth) {
      const zoom = desiredWidth > size.width - 34 ? (size.width - 34) / desiredWidth : 1
      return {
        width: Math.min(
          Math.max(50, Math.round(desiredWidth * (1 / zoom))),
          Math.round((size.width - 34) * (1 / zoom))
        ),
        zoom,
      }
    }

    function constrainHeight(desiredHeight) {
      const zoom = desiredHeight > size.height - 17 - 24 ? (size.height - 17 - 24) / desiredHeight : 1
      return {
        height: Math.min(
          Math.max(50, Math.round(desiredHeight * (1 / zoom))),
          Math.round((size.height - 17 - 24) * (1 / zoom))
        ),
        zoom,
      }
    }

    function constrainSize(desiredWidth, desiredHeight) {
      const { width, zoom: widthZoom } = constrainWidth(desiredWidth)
      const { height, zoom: heightZoom } = constrainHeight(desiredHeight)
      return {
        width,
        height,
        zoom: Math.min(widthZoom, heightZoom),
      }
    }

    useEffect(() => {
      let isInitial = true
      const observer = new ResizeObserver(() => {
        window.clearTimeout(timeout.current)
        const rect = containerRef.current.getBoundingClientRect()
        const width = Math.round(rect.width)
        const height = Math.round(rect.height)
        setSize({
          visible: !isInitial && width !== 0 && height !== 0,
          width,
          height,
        })
        timeout.current = window.setTimeout(() => {
          setSize((size) => ({ ...size, visible: false }))
        }, 1000)
        isInitial = false
      })
      observer.observe(containerRef.current)
      return () => {
        observer.disconnect()
      }
    }, [])

    useIsomorphicLayoutEffect(() => {
      if (size.width > 50 && size.height > 50) {
        onChangeResponsiveSize(({ width, height }) => ({ width, height }))
      }

      if (resizing) {
        function onMouseMove(e) {
          e.preventDefault()
          const { x, y } = getPointerPosition(e)
          if (resizing.handle === 'bottom') {
            document.body.classList.add('cursor-ns-resize')
            onChangeResponsiveSize(({ width }) => ({
              width,
              height: resizing.startHeight + (y - resizing.startY),
            }))
          } else if (resizing.handle === 'left') {
            document.body.classList.add('cursor-ew-resize')
            onChangeResponsiveSize(({ height }) => ({
              width: resizing.startWidth - (x - resizing.startX) * 2,
              height,
            }))
          } else if (resizing.handle === 'right') {
            document.body.classList.add('cursor-ew-resize')
            onChangeResponsiveSize(({ height }) => ({
              width: resizing.startWidth + (x - resizing.startX) * 2,
              height,
            }))
          } else if (resizing.handle === 'bottom-left') {
            document.body.classList.add('cursor-nesw-resize')
            onChangeResponsiveSize(() => ({
              width: resizing.startWidth - (x - resizing.startX) * 2,
              height: resizing.startHeight + (y - resizing.startY),
            }))
          } else if (resizing.handle === 'bottom-right') {
            document.body.classList.add('cursor-nwse-resize')
            onChangeResponsiveSize(() => ({
              width: resizing.startWidth + (x - resizing.startX) * 2,
              height: resizing.startHeight + (y - resizing.startY),
            }))
          }
        }
        function onMouseUp(e) {
          e.preventDefault()
          if (resizing.handle === 'bottom') {
            document.body.classList.remove('cursor-ns-resize')
          } else if (resizing.handle === 'left') {
            document.body.classList.remove('cursor-ew-resize')
          } else if (resizing.handle === 'right') {
            document.body.classList.remove('cursor-ew-resize')
          } else if (resizing.handle === 'bottom-left') {
            document.body.classList.remove('cursor-nesw-resize')
          } else if (resizing.handle === 'bottom-right') {
            document.body.classList.remove('cursor-nwse-resize')
          }
          setResizing()
        }
        window.addEventListener('mousemove', onMouseMove)
        window.addEventListener('mouseup', onMouseUp)
        window.addEventListener('touchmove', onMouseMove)
        window.addEventListener('touchend', onMouseUp)
        return () => {
          window.removeEventListener('mousemove', onMouseMove)
          window.removeEventListener('mouseup', onMouseUp)
          window.removeEventListener('touchmove', onMouseMove)
          window.removeEventListener('touchend', onMouseUp)
        }
      }
    }, [resizing, size])

    function startLeft(e) {
      const pos = getPointerPosition(e)
      if (pos === null) return
      e.preventDefault()
      setResizing({
        handle: 'left',
        startWidth: constrainedResponsiveSize.width,
        startX: pos.x,
      })
    }

    function startRight(e) {
      const pos = getPointerPosition(e)
      if (pos === null) return
      e.preventDefault()
      setResizing({
        handle: 'right',
        startWidth: constrainedResponsiveSize.width,
        startX: pos.x,
      })
    }

    function startBottomLeft(e) {
      const pos = getPointerPosition(e)
      if (pos === null) return
      e.preventDefault()
      setResizing({
        handle: 'bottom-left',
        startWidth: constrainedResponsiveSize.width,
        startHeight: constrainedResponsiveSize.height,
        startX: pos.x,
        startY: pos.y,
      })
    }

    function startBottom(e) {
      const pos = getPointerPosition(e)
      if (pos === null) return
      e.preventDefault()
      setResizing({
        handle: 'bottom',
        startHeight: constrainedResponsiveSize.height,
        startY: pos.y,
      })
    }

    function startBottomRight(e) {
      const pos = getPointerPosition(e)
      if (pos === null) return
      e.preventDefault()
      setResizing({
        handle: 'bottom-right',
        startWidth: constrainedResponsiveSize.width,
        startHeight: constrainedResponsiveSize.height,
        startX: pos.x,
        startY: pos.y,
      })
    }

    return (
      <div
        className="absolute inset-0 top-0 flex flex-col bg-black"
        ref={containerRef}
      >
        {responsiveDesignMode && (
          <ResponsiveSizing
            mediaScreen={mediaScreen}
            responsiveSize={constrainedResponsiveSize}
            onChangeResponsiveSize={onChangeResponsiveSize}
          />
        )}
        <div
          className="flex-auto grid justify-center"
          style={
            responsiveDesignMode
              ? {
                gridTemplateColumns: '1.0625rem min-content 1.0625rem',
                gridTemplateRows: 'min-content 1.0625rem',
              }
              : { gridTemplateColumns: '100%' }
          }
        >
          {responsiveDesignMode && (
            <div
              className="cursor-ew-resize select-none bg-gray-800 text-gray-600 hover:text-gray-400 transition-colors duration-150 flex items-center justify-center"
              onMouseDown={startLeft}
              onTouchStart={startLeft}
            >
              <svg
                className="w-1.5 h-8"
                viewBox="0 0 6 32"
                fill="none"
                stroke="currentColor"
              >
                <path d="M 0.5 0 V 32 M 5.5 0 V 32" />
              </svg>
            </div>
          )}
          <div
            className={`relative${responsiveDesignMode ? ' border border-gray-700 shadow-sm overflow-hidden' : ''}`}
            style={
              responsiveDesignMode
                ? {
                  width: Math.round(
                    constrainedResponsiveSize.width *
                    constrainedResponsiveSize.zoom
                  ),
                  height: Math.round(
                    constrainedResponsiveSize.height *
                    constrainedResponsiveSize.zoom
                  ),
                }
                : {}
            }
          >
            <iframe
              ref={ref}
              title="Preview"
              id="frame"
              name="frame"
              src={frameUrl}
              style={
                responsiveDesignMode
                  ? {
                    width: constrainedResponsiveSize.width,
                    height: constrainedResponsiveSize.height,
                    marginLeft:
                      (constrainedResponsiveSize.width -
                        Math.round(
                          constrainedResponsiveSize.width *
                          constrainedResponsiveSize.zoom
                        )) /
                      -2,
                    transformOrigin: 'top',
                    transform: `scale(${constrainedResponsiveSize.zoom})`,
                  }
                  : {}
              }
              className={`absolute inset-0 w-full h-full bg-white ${resizing ? 'pointer-events-none select-none' : ''}`}
            />
          </div>
          {responsiveDesignMode && (
            <>
              <div
                className="cursor-ew-resize select-none bg-gray-800 text-gray-600 hover:text-gray-400 transition-colors duration-150 flex items-center justify-center"
                onMouseDown={startRight}
                onTouchStart={startRight}
              >
                <svg
                  className="w-1.5 h-8"
                  viewBox="0 0 6 32"
                  fill="none"
                  stroke="currentColor"
                >
                  <path d="M 0.5 0 V 32 M 5.5 0 V 32" />
                </svg>
              </div>
              <div
                className="cursor-nesw-resize select-none bg-gray-800 text-gray-600 hover:text-gray-400 transition-colors duration-150 flex items-center justify-center"
                onMouseDown={startBottomLeft}
                onTouchStart={startBottomLeft}
              >
                <svg
                  viewBox="0 0 16 6"
                  width={16}
                  height={6}
                  fill="none"
                  stroke="currentColor"
                  className="transform translate-x-0.5 -translate-y-0.5 rotate-45"
                >
                  <path d="M 0 0.5 H 16 M 0 5.5 H 16" />
                </svg>
              </div>
              <div
                className="cursor-ns-resize select-none bg-gray-800 text-gray-600 hover:text-gray-400 transition-colors duration-150 flex items-center justify-center"
                onMouseDown={startBottom}
                onTouchStart={startBottom}
              >
                <svg
                  className="w-8 h-1.5"
                  viewBox="0 0 32 6"
                  fill="none"
                  stroke="currentColor"
                >
                  <path d="M 0 0.5 H 32 M 0 5.5 H 32" />
                </svg>
              </div>
              <div
                className="cursor-nwse-resize select-none bg-gray-800 text-gray-600 hover:text-gray-400 transition-colors duration-150 flex items-center justify-center"
                onMouseDown={startBottomRight}
                onTouchStart={startBottomRight}
              >
                <svg
                  viewBox="0 0 16 6"
                  width={16}
                  height={6}
                  fill="none"
                  stroke="currentColor"
                  className="transform -translate-x-0.5 -translate-y-0.5 -rotate-45"
                >
                  <path d="M 0 0.5 H 16 M 0 5.5 H 16" />
                </svg>
              </div>
            </>
          )}
        </div>
        {
          !responsiveDesignMode && size.visible && (
            <div className="absolute top-0.5 mx-auto w-full flex justify-center z-[101]">
              <span className="flex content-center rounded-full text-xs whitespace-pre px-2 py-0 tabular-nums border border-gray-400/50 shadow bg-gray-600/80 select-none">
                {size.width}
                {'  '}×{'  '}
                {size.height}
              </span>
            </div>
          )
        }
      </div >
    )
  }
)


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
const vscode = require("vscode");
function activate(context) {
    const onOpen = vscode.commands.registerCommand('preview.open', () => {
        PreviewPanel.createOrShow(context.extensionUri);
    });
    const onUrl = vscode.commands.registerCommand('preview.url', async () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.sendUrl();
        }
    });
    const onBack = vscode.commands.registerCommand('preview.back', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.sendPageBack();
        }
    });
    const onForward = vscode.commands.registerCommand('preview.forward', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.sendPageForward();
        }
    });
    const onRefresh = vscode.commands.registerCommand('preview.refresh', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.sendRefresh();
        }
    });
    const onResponsiveView = vscode.commands.registerCommand('preview.responsiveView', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.sendResponsiveView();
        }
    });
    const onScreenView = vscode.commands.registerCommand('preview.screenView', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.toggleScreenView();
        }
    });
    const onOpenDevTools = vscode.commands.registerCommand('preview.openDevTools', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.openDevTools();
        }
    });
    const onOpenInBrowser = vscode.commands.registerCommand('preview.openInBrowser', () => {
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel.openInBrowser();
        }
    });
    context.subscriptions.push(onOpen, onUrl, onBack, onForward, onRefresh, onResponsiveView, onScreenView, onOpenDevTools, onOpenInBrowser);
    if (vscode.window.registerWebviewPanelSerializer) {
        // Make sure we register a serializer in activation event
        vscode.window.registerWebviewPanelSerializer(PreviewPanel.viewType, {
            async deserializeWebviewPanel(webviewPanel, state) {
                // Reset the webview options so we use latest uri for `localResourceRoots`.
                webviewPanel.webview.options = getWebviewOptions(context.extensionUri);
                PreviewPanel.revive(webviewPanel, context.extensionUri, state?.previewUrl);
            }
        });
    }
}
exports.activate = activate;
function getWebviewOptions(extensionUri) {
    return {
        // Enable javascript in the webview
        enableScripts: true,
        // And restrict the webview to only loading content from our extension's `media` directory.
        localResourceRoots: [vscode.Uri.joinPath(extensionUri, 'media')]
    };
}
/**
 * Manages Preview webview panels
 */
class PreviewPanel {
    static createOrShow(extensionUri) {
        // If we already have a panel, show it.
        if (PreviewPanel.currentPanel) {
            PreviewPanel.currentPanel._panel.reveal(PreviewPanel.currentPanel._panel.viewColumn);
            return;
        }
        // Otherwise, create a new panel.
        const panel = vscode.window.createWebviewPanel(PreviewPanel.viewType, 'Preview', vscode.ViewColumn.Two, getWebviewOptions(extensionUri));
        PreviewPanel.currentPanel = new PreviewPanel(panel, extensionUri);
    }
    static revive(panel, extensionUri, oldUrl) {
        PreviewPanel.currentPanel = new PreviewPanel(panel, extensionUri);
        PreviewPanel.currentPanel._currentUrl = oldUrl;
    }
    constructor(panel, extensionUri) {
        this._disposables = [];
        this._panel = panel;
        this._extensionUri = extensionUri;
        vscode.commands.executeCommand('setContext', 'PreviewActive', true);
        this._panel.webview.html = this._getHtmlForWebview(this._panel.webview);
        panel.iconPath = {
            light: vscode.Uri.joinPath(extensionUri, 'media', 'icons', 'preview-light.svg'),
            dark: vscode.Uri.joinPath(extensionUri, 'media', 'icons', 'preview-dark.svg'),
        };
        // Set the webview's initial html content
        this._update();
        // Listen for when the panel is disposed
        // This happens when the user closes the panel or when the panel is closed programmatically
        this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
        // Update the content based on view changes
        this._panel.onDidChangeViewState(e => {
            if (this._panel.visible) {
                this._update();
            }
        }, null, this._disposables);
        vscode.workspace.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration('webPreview')) {
                this._currentUrl = undefined;
                this._update();
            }
        });
        // Handle messages from the webview
        this._panel.webview.onDidReceiveMessage(message => {
            if (message.command) {
                vscode.window.showInformationMessage(message.text);
                return;
            }
        }, null, this._disposables);
    }
    async sendUrl() {
        const url = await vscode.window.showInputBox({
            value: this._currentUrl,
            // @ts-ignore
            valueSelection: [this._currentUrl?.length, this._currentUrl?.length],
            placeHolder: 'http://localhost:3000'
        });
        if (!url)
            return;
        this._currentUrl = url;
        this._panel.webview.postMessage({ preview: { url } });
    }
    sendPageBack() {
        this._panel.webview.postMessage({ preview: { back: true } });
    }
    sendPageForward() {
        this._panel.webview.postMessage({ preview: { forward: true } });
    }
    sendRefresh() {
        this._panel.webview.postMessage({ preview: { refresh: true } });
    }
    sendResponsiveView() {
        this._panel.webview.postMessage({ preview: { responsive: true } });
    }
    toggleScreenView() {
        // @ts-ignore
        if (this._panel.viewColumn > 1) {
            this._panel.reveal(vscode.ViewColumn.One);
        }
        else if (this._panel.viewColumn == 1) {
            this._panel.reveal(vscode.ViewColumn.Two);
        }
    }
    openDevTools() {
        vscode.commands.executeCommand('workbench.action.webview.openDeveloperTools');
    }
    openInBrowser() {
        const url = vscode.Uri.parse(`${this._currentUrl}`);
        vscode.env.openExternal(url);
    }
    dispose() {
        PreviewPanel.currentPanel = undefined;
        vscode.commands.executeCommand('setContext', 'PreviewActive', false);
        // Clean up our resources
        this._panel.dispose();
        while (this._disposables.length) {
            const x = this._disposables.pop();
            if (x) {
                x.dispose();
            }
        }
    }
    _update() {
        const config = vscode.workspace.getConfiguration('webPreview');
        const url = config.get('url', 'http://localhost:3000');
        if (!this._currentUrl) {
            this._currentUrl = url;
        }
        const mediaScreenOverride = config.get('mediaScreenOverride', false);
        let mediaScreen = config.get('mediaScreen', {});
        if (mediaScreenOverride) {
            const inspect = config.inspect('mediaScreen');
            // @ts-ignore
            mediaScreen = inspect.workspaceValue ? inspect.workspaceValue : inspect.globalValue ? inspect.globalValue : inspect.defaultValue;
        }
        // Vary the webview's content based on where it is located in the editor.
        if (this._panel.viewColumn) {
            this._panel.webview.postMessage({ preview: { url: this._currentUrl, mediaScreen } });
        }
    }
    _getHtmlForWebview(webview) {
        // Local path to main script run in the webview
        const scriptPathOnDisk = vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js');
        // And the uri we use to load this script in the webview
        const scriptUri = webview.asWebviewUri(scriptPathOnDisk);
        // Local path to css styles
        const stylesPathMainPath = vscode.Uri.joinPath(this._extensionUri, 'media', 'main.css');
        // Uri to load styles into webview
        const stylesMainUri = webview.asWebviewUri(stylesPathMainPath);
        // Use a nonce to only allow specific scripts to be run
        const nonce = getNonce();
        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
                <meta name="theme-color" content="#000000">
				<title>Preview</title>
				<link href="${stylesMainUri}" rel="stylesheet">
                <script defer="defer" nonce="${nonce}" src="${scriptUri}"></script>
			</head>
			<body class="w-full min-h-full flex text-white bg-gray-900 !p-0 overflow-hidden">
                <noscript>You need to enable JavaScript to run this app.</noscript>
                <div id="root" class="w-full flex flex-none flex-col"></div>
            </body>
			</html>`;
    }
}
PreviewPanel.viewType = 'preview';
function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}
//# sourceMappingURL=extension.js.map